client/client-lib/src/mint/mod.rs

```diff
@@ -222,43 +224,44 @@ impl<'c> MintClient<'c> {
 
-    pub fn list_active_issuances(&self) -> Vec<OutPoint> {
+    pub fn list_active_issuances(&self) -> Vec<(OutPoint, CoinFinalizationData)> {
         self.context
             .db
             .find_by_prefix(&OutputFinalizationKeyPrefix)
-            .map(|res| res.expect("DB error").0 .0)
+            .map(|res| {
+                let (OutputFinalizationKey(outpoint), cfd) = res.expect("DB error");
+                (outpoint, cfd)
+            })
             .collect()
     }
 
-    // FIXME: remove
-    pub async fn fetch_all_coins(&self, mut batch: BatchTx<'_>) -> Result<Vec<TransactionId>> {
-        let active_issuances = self
-            .context
-            .db
-            .find_by_prefix(&OutputFinalizationKeyPrefix)
-            .collect::<std::result::Result<Vec<_>, _>>()
-            .expect("DB error");
-
-        // TODO: return out points instead
-        let mut tx_ids = vec![];
-        for (OutputFinalizationKey(out_point), _) in active_issuances {
-            loop {
-                match self.fetch_coins(batch.subtransaction(), out_point).await {
-                    Ok(()) => {
-                        tx_ids.push(out_point.txid);
-                        break;
-                    }
-                    // TODO: make mint error more expressive (currently any HTTP error) and maybe use custom return type instead of error for retrying
-                    Err(e) if e.is_retryable() => {
-                        trace!("Mint returned retryable error: {:?}", e);
-                        minimint_api::task::sleep(Duration::from_secs(1)).await
+    pub async fn fetch_all_coins(&self) -> Result<Vec<OutPoint>> {
+        let active_issuances = self.list_active_issuances();
+        if active_issuances.is_empty() {
+            return Ok(Vec::new());
+        }
+
+        let stream = active_issuances
+            .into_iter()
+            .map(|(out_point, _)| async move {
+                let mut batch = DbBatch::new();
+                loop {
+                    match self.fetch_coins(batch.transaction(), out_point).await {
+                        Ok(_) => {
+                            self.context.db.apply_batch(batch).expect("DB error");
+                            return Ok(out_point);
+                        }
+                        // TODO: make mint error more expressive (currently any HTTP error) and maybe use custom return type instead of error for retrying
+                        Err(e) if e.is_retryable() => {
+                            trace!("Mint returned retryable error: {:?}", e);
+                            minimint_api::task::sleep(Duration::from_secs(1)).await
+                        }
+                        Err(e) => return Err(e),
                     }
-                    Err(e) => return Err(e),
                 }
-            }
-        }
-        batch.commit();
-        Ok(tx_ids)
+            })
+            .collect::<FuturesUnordered<_>>();
+        stream.try_collect().await
```

Comment:
I guess one issue here is that because the write to db happens inside the stream, it's possible this will mutate the db and return an Err without returning the OutPoints that were deleted.

I don't think anything currently relies on the return value of this function for getting all the fetched OutPoints, but it could be an issue if clients do.

Reply:
Right, we might actually want to return a `Vec<Result<_, _>>`.